examples/sampling/graphbolt/temporal_link_prediction.py

```diff
@@ -0,0 +1,367 @@
+"""
+This script trains and tests a Heterogeneous GraphSAGE model for link
+prediction with temporal information using graphbolt dataloader.
+
+Unlike previous dgl examples, we've utilized the newly defined dataloader
+from GraphBolt. This example will help you grasp how to build an end-to-end
+training pipeline using GraphBolt.
+
+While node classification predicts labels for nodes based on their
+local neighborhoods, link prediction assesses the likelihood of an edge
+existing between two nodes, necessitating different sampling strategies
+that account for pairs of nodes and their joint neighborhoods.
+
+This flowchart describes the main functional sequence of the provided example.
+main
+│
+├───> OnDiskDataset pre-processing
+│
+├───> Instantiate HeteroSAGE model
+│
+├───> train
+│     │
+│     ├───> Get graphbolt dataloader (HIGHLIGHT)
+│     │
+│     └───> Training loop
+│           │
+│           ├───> HeteroSAGE.forward
+│           │
+│           └───> Validation set evaluation
+│
+└───> Test set evaluation
+"""
+import argparse
+import os
+import time
+from functools import partial
+
+import dgl.graphbolt as gb
+import dgl.nn as dglnn
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import tqdm
+from dgl.data.utils import download, extract_archive
+
+
+TIMESTAMP_FEATURE_NAME = "__timestamp__"
+node_feature_keys = {
+    "Product": ["categoryId"],
+    "Query": ["categoryId"],
+}
+
+target_type = ("Query", "Click", "Product")
+all_connected_types = [
+    target_type,
+    ("Product", "reverse_Click", "Query"),
+    ("Product", "reverse_QueryResult", "Query"),
+    ("Query", "QueryResult", "Product"),
+]
+
+
+class CategoricalEncoder(nn.Module):
+    def __init__(
+        self,
+        num_categories,
+        out_size,
+    ):
+        super().__init__()
+        self.embed = nn.Embedding(num_categories, out_size)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.embed.weight)
+
+    def forward(self, input_feat: torch.Tensor):
+        return self.embed(input_feat.view(-1))
+
+
+class HeteroSAGE(nn.Module):
+    def __init__(self, in_size, hidden_size):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        self.layers.append(
+            dglnn.HeteroGraphConv(
```

Comment:
Can we add these two layers in a loop? It looks like only the dimensions are changing.